File: src/ui/widgets/song_row.py
import gi

gi.require_version("Gtk", "4.0")
gi.require_version("Adw", "1")
from gi.repository import Gtk, Adw, GLib, Pango
from ui.utils import AsyncPicture, LikeButton, show_toast
from ui.context_menu import show_song_menu

# AlbumPage is resolved lazily because ui.pages.album imports BasePlaylistPage,
# which transitively imports this module — a top-level import would cycle.
# Caching the class reference avoids re-importing inside every bind().
_ALBUM_PAGE_CLS = None


def _is_album_page(page):
    global _ALBUM_PAGE_CLS
    if _ALBUM_PAGE_CLS is None:
        from ui.pages.album import AlbumPage
        _ALBUM_PAGE_CLS = AlbumPage
    return isinstance(page, _ALBUM_PAGE_CLS)


class SongRowWidget(Gtk.Box):
    def __init__(self, player, client):
        super().__init__(orientation=Gtk.Orientation.HORIZONTAL)
        self.player = player
        self.client = client
        self.model_item = None
        self._notify_handler_id = None
        self._player_handler_id = None
        self._start_x = 0
        self._start_y = 0

        # Keep the downloaded indicator in sync across any view showing this
        # song — react to both completions and removals from elsewhere in the UI.
        dm = self.player.download_manager
        self._dm_done_handler = dm.connect("item-done", self._on_dm_item_done)
        self._dm_removed_handler = dm.connect("download-removed", self._on_dm_download_removed)
        self.connect("destroy", self._on_destroy)

        self.row = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        self.row.set_hexpand(True)
        self.row.add_css_class("song-row")
        self.append(self.row)

        # Image with playing indicator overlay
        self.img = AsyncPicture(crop_to_square=True, target_size=56, player=self.player)
        self.img.add_css_class("song-img")

        self.img_overlay = Gtk.Overlay()
        self.img_overlay.set_child(self.img)
        self.img_overlay.set_valign(Gtk.Align.CENTER)

        # Track number label (for album view)
        self.track_num_label = Gtk.Label()
        self.track_num_label.add_css_class("dim-label")
        self.track_num_label.add_css_class("caption")
        self.track_num_label.set_valign(Gtk.Align.CENTER)
        self.track_num_label.set_halign(Gtk.Align.CENTER)
        self.track_num_label.set_size_request(40, 40)
        self.track_num_label.set_visible(False)

        # Playing indicator: 3 animated bars
        self.playing_indicator = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
        self.playing_indicator.set_halign(Gtk.Align.CENTER)
        self.playing_indicator.set_valign(Gtk.Align.CENTER)
        self.playing_indicator.add_css_class("playing-indicator")
        self.playing_indicator.set_visible(False)

        self.bar1 = Gtk.Box()
        self.bar1.add_css_class("playing-bar")
        self.bar1.add_css_class("playing-bar-1")
        self.bar2 = Gtk.Box()
        self.bar2.add_css_class("playing-bar")
        self.bar2.add_css_class("playing-bar-2")
        self.bar3 = Gtk.Box()
        self.bar3.add_css_class("playing-bar")
        self.bar3.add_css_class("playing-bar-3")

        self.playing_indicator.append(self.bar1)
        self.playing_indicator.append(self.bar2)
        self.playing_indicator.append(self.bar3)

        self._anim_timer_id = None
        self._anim_state = False

        self.img_overlay.add_overlay(self.playing_indicator)
        self.row.append(self.track_num_label)
        self.row.append(self.img_overlay)

        # Main Title / Subtitle Box
        self.vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)
        self.vbox.set_valign(Gtk.Align.CENTER)
        self.vbox.set_hexpand(True)

        # Title + badges live in `title_box`. We want the badges to sit
        # directly next to the title text, not at the far right of the row.
        # `title_label` has hexpand=False (so it doesn't gobble all leftover
        # space and push the badges away) but width_chars=1 + ellipsize=END
        # so it can still shrink when the row is narrow. The trailing
        # `_lv_title_spacer` is the element that absorbs leftover space,
        # keeping the trio left-aligned within title_box.
        self.title_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
        self.title_label = Gtk.Label()
        self.title_label.set_halign(Gtk.Align.START)
        self.title_label.set_xalign(0.0)
        self.title_label.set_hexpand(False)
        self.title_label.set_ellipsize(Pango.EllipsizeMode.END)
        self.title_label.set_lines(1)
        self.title_label.set_width_chars(1)

        self.explicit_badge = Gtk.Label(label="E")
        self.explicit_badge.add_css_class("explicit-badge")
        self.explicit_badge.set_valign(Gtk.Align.CENTER)
        self.explicit_badge.set_halign(Gtk.Align.CENTER)
        self.explicit_badge.set_justify(Gtk.Justification.CENTER)
        self.explicit_badge.set_visible(False)

        self.dl_icon = Gtk.Image.new_from_icon_name("folder-download-symbolic")
        self.dl_icon.set_pixel_size(14)
        self.dl_icon.add_css_class("dim-label")
        self.dl_icon.set_valign(Gtk.Align.CENTER)
        self.dl_icon.set_visible(False)

        self.title_box.append(self.title_label)
        self.title_box.append(self.explicit_badge)
        self.title_box.append(self.dl_icon)

        # Soak up any leftover horizontal space so the title + badges stay
        # packed at the start of the row.
        title_spacer = Gtk.Box()
        title_spacer.set_hexpand(True)
        self.title_box.append(title_spacer)

        self.subtitle_label = Gtk.Label()
        self.subtitle_label.set_halign(Gtk.Align.START)
        self.subtitle_label.set_ellipsize(Pango.EllipsizeMode.END)
        self.subtitle_label.set_lines(1)
        self.subtitle_label.add_css_class("dim-label")
        self.subtitle_label.add_css_class("caption")

        self.vbox.append(self.title_box)
        self.vbox.append(self.subtitle_label)
        self.row.append(self.vbox)

        # Suffixes: Duration, Like

        self.dur_lbl = Gtk.Label()
        self.dur_lbl.add_css_class("caption")
        self.dur_lbl.set_valign(Gtk.Align.CENTER)
        self.dur_lbl.set_margin_end(6)
        self.row.append(self.dur_lbl)

        self.like_btn = LikeButton(self.client, None)
        self.like_btn.set_valign(Gtk.Align.CENTER)
        self.row.append(self.like_btn)

        # Gesture for Right Click (Context Menu)
        gesture = Gtk.GestureClick()
        gesture.set_button(3)  # Right click
        gesture.connect("released", self.on_right_click)
        self.row.add_controller(gesture)

        # Long Press for touch
        lp = Gtk.GestureLongPress()
        lp.connect("pressed", lambda g, x, y: self.on_right_click(g, 1, x, y))
        self.row.add_controller(lp)

        # Gesture for Left Click (Activation)
        left_click = Gtk.GestureClick()
        left_click.set_button(1)
        left_click.connect("pressed", self._on_left_pressed)
        left_click.connect("released", self._on_left_released)
        self.row.add_controller(left_click)

    def _current_video_id(self):
        return self.model_item.video_id if self.model_item else None

    def _on_dm_item_done(self, dm, video_id, success, message):
        if success and video_id and video_id == self._current_video_id():
            self.dl_icon.set_visible(True)

    def _on_dm_download_removed(self, dm, video_id):
        if video_id and video_id == self._current_video_id():
            self.dl_icon.set_visible(False)

    def _on_destroy(self, widget):
        dm = self.player.download_manager
        for hid_attr in ("_dm_done_handler", "_dm_removed_handler"):
            hid = getattr(self, hid_attr, None)
            if hid is not None:
                try:
                    dm.disconnect(hid)
                except Exception:
                    pass
                setattr(self, hid_attr, None)

    def bind(self, item, page):
        # Disconnect previous player signal handler
        if self._player_handler_id is not None:
            self.player.disconnect(self._player_handler_id)
            self._player_handler_id = None
        # Disconnect previous item notify handler
        if self._notify_handler_id is not None and self.model_item is not None:
            try:
                self.model_item.disconnect(self._notify_handler_id)
            except Exception:
                pass
            self._notify_handler_id = None

        self.model_item = item
        self.page = page

        self.title_label.set_label(item.title)
        self.title_label.set_tooltip_text(item.title)
        self.subtitle_label.set_label(item.artist)
        self.subtitle_label.set_tooltip_text(item.artist)

        self.dur_lbl.set_label(item.duration)
        self.explicit_badge.set_visible(item.is_explicit)

        # Check if this is an album view
        is_album = _is_album_page(page)

        if is_album:
            # Show track number instead of thumbnail
            self.track_num_label.set_label(str(item.index + 1))
            self.track_num_label.set_visible(True)
            self.img_overlay.set_visible(False)
        else:
            self.track_num_label.set_visible(False)
            self.img_overlay.set_visible(True)
            self.img.video_id = item.video_id
            self.img.load_url(item.thumbnail_url)

        self.like_btn.set_data(item.video_id, item.like_status)

        # Downloaded indicator
        if item.video_id:
            self.dl_icon.set_visible(self.player.download_manager.is_downloaded(item.video_id))
        else:
            self.dl_icon.set_visible(False)

        if not item.video_id:
            self.row.set_sensitive(False)
        else:
            self.row.set_sensitive(True)

        # CSS handles responsiveness and size limits natively now

        # Set initial playing state. Check both the player's current
        # video id AND the swap-source id so the row lights up even
        # when the player has already swapped the queued OMV track to
        # its ATV counterpart before this bind happened.
        _source_vid = getattr(self.player, "_current_source_video_id", None)
        self._apply_playing_state(
            bool(item.video_id and (
                item.video_id == self.player.current_video_id
                or item.video_id == _source_vid
            ))
        )

        # Connect directly to the player metadata signal (reliable than GObject property notify)
        self._player_handler_id = self.player.connect(
            "metadata-changed", self._on_player_metadata_changed
        )

    def _on_player_metadata_changed(self, player, *args):
        if self.model_item:
            vid = self.model_item.video_id
            # Match the swap-from id too: when the player auto-swaps
            # an OMV/UGC track to its audio (ATV) counterpart, the
            # `current_video_id` drifts away from what an album/
            # playlist row holds. `_current_source_video_id` is the
            # pre-swap id, so the highlight stays on the right row.
            source_vid = getattr(player, "_current_source_video_id", None)
            is_playing = bool(
                vid and (vid == player.current_video_id or vid == source_vid)
            )
            self._apply_playing_state(is_playing)

    def stop_handlers(self):
        """Disconnect all signal handlers. Called on factory unbind."""
        if self._player_handler_id is not None:
            try:
                self.player.disconnect(self._player_handler_id)
            except Exception:
                pass
            self._player_handler_id = None
        if self._notify_handler_id is not None and self.model_item is not None:
            try:
                self.model_item.disconnect(self._notify_handler_id)
            except Exception:
                pass
            self._notify_handler_id = None
        self._stop_animation()
        # Release the thumbnail texture so GTK can free it — the widget
        # may be recycled for a different track with a different cover,
        # and leaving the old paintable attached keeps the texture ref
        # alive until the bind for the new track completes. Doing this
        # explicitly bounds peak memory when scrolling through long lists.
        try:
            if hasattr(self, "img") and self.img is not None:
                self.img.set_paintable(None)
                self.img.url = None
                self.img._is_placeholder = True
        except Exception:
            pass

    def _apply_playing_state(self, is_playing):
        if is_playing:
            self.row.add_css_class("playing")
            self.playing_indicator.set_visible(True)
            self._start_animation()
        else:
            self.row.remove_css_class("playing")
            self.playing_indicator.set_visible(False)
            self._stop_animation()

    def _start_animation(self):
        if self._anim_timer_id is not None:
            return  # Already running
        self._anim_state = False
        self._anim_timer_id = GLib.timeout_add(350, self._tick_animation)

    def _stop_animation(self):
        if self._anim_timer_id is not None:
            GLib.source_remove(self._anim_timer_id)
            self._anim_timer_id = None
        # Reset bars to default state
        self.bar1.remove_css_class("bar-up")
        self.bar2.remove_css_class("bar-up")
        self.bar3.remove_css_class("bar-up")

    def _tick_animation(self):
        self._anim_state = not self._anim_state
        if self._anim_state:
            self.bar1.add_css_class("bar-up")
            self.bar3.add_css_class("bar-up")
            self.bar2.remove_css_class("bar-up")
        else:
            self.bar2.add_css_class("bar-up")
            self.bar1.remove_css_class("bar-up")
            self.bar3.remove_css_class("bar-up")
        return GLib.SOURCE_CONTINUE

    def _on_left_pressed(self, gesture, n_press, x, y):
        self._start_x = x
        self._start_y = y

    def _on_left_released(self, gesture, n_press, x, y):
        # Displacement check
        dx = abs(x - self._start_x)
        dy = abs(y - self._start_y)
        if dx > 10 or dy > 10:
            return

        if self.model_item and self.page:
            # Trigger page activation logic
            if hasattr(self.page, "on_song_activated"):
                # Pass the SongItem itself, not its stored .index — that index
                # is the original position in the underlying store and goes
                # stale once the user sorts or filters the view, which would
                # send the wrong position to the player.
                self.page.on_song_activated(None, self.model_item)

    def on_right_click(self, gesture, n_press, x, y):
        if not self.model_item:
            return

        item = self.model_item
        show_song_menu(
            self.row,
            x,
            y,
            item.track_data,
            player=self.player,
            client=self.client,
            prefix="row",
            video_id=item.video_id,
            album_title=getattr(self.page, "playlist_title_text", None),
            album_id=getattr(self.page, "playlist_id", None),
        )

    def _show_toast(self, message):
        show_toast(self, message)

File: src/ui/widgets/song_row.rs
//! Port of ui/widgets/song_row.py: thumbnail with a playing indicator, title
//! with explicit and downloaded badges, subtitle, duration, like button,
//! right-click menu, and a click that ignores drags.

use std::cell::{Cell, RefCell};
use std::rc::Rc;
use std::time::Duration;

use gtk::{gdk, glib, prelude::*};

use crate::model::{ItemKind, MediaItem, VideoId};
use crate::ui::context::UiContext;
use crate::ui::context_menu::{MenuAction, SongMenuOptions, show_song_menu};
use crate::ui::cover::CoverImage;
use crate::ui::like_button::LikeButton;
use crate::ui::widgets::song_list::{kind_subtitle, kind_subtitle_text, search_subtitle};

const THUMB_SIZE: i32 = 56;
const CLICK_SLOP: f64 = 10.0;
const ANIMATION_STEP: Duration = Duration::from_millis(350);

type ActivateHandler = Rc<dyn Fn(&MediaItem)>;
/// What a page adds to this row's menu, built fresh each time it opens.
type MenuExtras = Rc<dyn Fn() -> Vec<MenuAction>>;

pub struct SongRow {
    row: gtk::ListBoxRow,
    inner: gtk::Box,
    cover: Rc<CoverImage>,
    img_overlay: gtk::Overlay,
    track_num: gtk::Label,
    indicator: gtk::Box,
    bars: [gtk::Box; 3],
    title: gtk::Label,
    explicit: gtk::Label,
    dl_icon: gtk::Image,
    subtitle_box: gtk::Box,
    duration: gtk::Label,
    like: Rc<LikeButton>,
    ctx: Rc<UiContext>,
    item: RefCell<Option<MediaItem>>,
    anim: RefCell<Option<glib::SourceId>>,
    anim_state: Cell<bool>,
    state_handler: RefCell<Option<glib::SignalHandlerId>>,
    on_activate: RefCell<Option<ActivateHandler>>,
    press_at: Cell<(f64, f64)>,
    /// Search rows tint the ListBoxRow and show no indicator, like attach_playing_highlight.
    row_highlight: Cell<bool>,
    /// A subtitle the page named itself, drawn plain with no kind icon.
    plain_subtitle: RefCell<Option<String>>,
    /// Extra menu entries the page contributes, built when the menu opens.
    menu_extras: RefCell<Option<MenuExtras>>,
}

impl SongRow {
    pub fn new(ctx: Rc<UiContext>) -> Rc<Self> {
        let row = gtk::ListBoxRow::builder().activatable(true).build();
        let inner = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).spacing(12).hexpand(true).css_classes(["song-row"]).build();
        row.set_child(Some(&inner));

        let cover = CoverImage::in_context(&ctx, THUMB_SIZE);
        cover.widget().add_css_class("song-img");
        let img_overlay = gtk::Overlay::builder().valign(gtk::Align::Center).child(cover.widget()).build();

        let track_num = gtk::Label::builder().css_classes(["dim-label", "caption"]).valign(gtk::Align::Center).halign(gtk::Align::Center).width_request(40).height_request(40).visible(false).build();

        // Same square and radius as the thumbnail, bars centered near the bottom.
        let indicator = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).halign(gtk::Align::Center).valign(gtk::Align::Center).css_classes(["playing-indicator", "song-img"]).visible(false).build();
        let bars_box = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).spacing(3).hexpand(true).halign(gtk::Align::Center).valign(gtk::Align::End).margin_bottom(12).build();
        let bars = [1, 2, 3].map(|n| {
            let bar = gtk::Box::builder().valign(gtk::Align::End).css_classes(["playing-bar", &format!("playing-bar-{n}")]).build();
            bars_box.append(&bar);
            bar
        });
        indicator.append(&bars_box);
        img_overlay.add_overlay(&indicator);
        inner.append(&track_num);
        inner.append(&img_overlay);

        let text = gtk::Box::builder().orientation(gtk::Orientation::Vertical).spacing(2).valign(gtk::Align::Center).hexpand(true).build();
        let title_box = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).spacing(6).build();
        let title = gtk::Label::builder().halign(gtk::Align::Start).xalign(0.0).hexpand(false).ellipsize(gtk::pango::EllipsizeMode::End).lines(1).width_chars(1).build();
        let explicit = gtk::Label::builder().label("E").css_classes(["explicit-badge"]).valign(gtk::Align::Center).halign(gtk::Align::Center).visible(false).build();
        let dl_icon = gtk::Image::builder().icon_name("folder-download-symbolic").pixel_size(14).css_classes(["dim-label"]).valign(gtk::Align::Center).visible(false).build();
        title_box.append(&title);
        title_box.append(&explicit);
        title_box.append(&dl_icon);
        title_box.append(&gtk::Box::builder().hexpand(true).build());
        let subtitle_box = gtk::Box::builder().orientation(gtk::Orientation::Horizontal).build();
        text.append(&title_box);
        text.append(&subtitle_box);
        inner.append(&text);

        let duration = gtk::Label::builder().css_classes(["caption"]).valign(gtk::Align::Center).margin_end(6).build();
        inner.append(&duration);
        let like = LikeButton::new(ctx.player.clone());
        like.widget().set_valign(gtk::Align::Center);
        inner.append(like.widget());

        let this = Rc::new(Self {
            row,
            inner,
            cover,
            img_overlay,
            track_num,
            indicator,
            bars,
            title,
            explicit,
            dl_icon,
            subtitle_box,
            duration,
            like,
            ctx,
            item: RefCell::new(None),
            anim: RefCell::new(None),
            anim_state: Cell::new(false),
            state_handler: RefCell::new(None),
            on_activate: RefCell::new(None),
            press_at: Cell::new((0.0, 0.0)),
            row_highlight: Cell::new(false),
            plain_subtitle: RefCell::new(None),
            menu_extras: RefCell::new(None),
        });
        this.connect_gestures();
        this
    }

    pub fn widget(&self) -> &gtk::ListBoxRow {
        &self.row
    }

    /// Search-page row: tint the outer ListBoxRow, no indicator, no duration column,
    /// and the "Kind • Artist • Album" subtitle search.py built.
    pub fn set_search_style(&self, enabled: bool) {
        self.row_highlight.set(enabled);
        self.duration.set_visible(!enabled);
    }

    /// Entries this row's menu carries on top of the standard ones, the way
    /// the history page adds Play and Remove from History.
    pub fn set_menu_extras(&self, f: impl Fn() -> Vec<MenuAction> + 'static) {
        self.menu_extras.replace(Some(Rc::new(f)));
    }

    /// Draw this text under the title instead of the kind line, the way the
    /// category page and the history page write their own.
    pub fn set_plain_subtitle(&self, text: Option<String>) {
        self.plain_subtitle.replace(text);
    }

    /// Without this the row relies on its ListBox's row-activated signal.
    #[allow(dead_code)]
    pub fn set_on_activate(&self, f: impl Fn(&MediaItem) + 'static) {
        self.on_activate.replace(Some(Rc::new(f)));
    }

    /// Fill the row. A track number replaces the thumbnail on album pages.
    pub fn bind(self: &Rc<Self>, item: &MediaItem, track_number: Option<u32>) {
        self.item.replace(Some(item.clone()));
        self.title.set_label(&item.title);
        self.title.set_tooltip_text(Some(&item.title));
        while let Some(child) = self.subtitle_box.first_child() {
            self.subtitle_box.remove(&child);
        }
        if let Some(text) = self.plain_subtitle.borrow().clone() {
            if !text.is_empty() {
                self.subtitle_box.append(&gtk::Label::builder().label(&text).halign(gtk::Align::Start).xalign(0.0).ellipsize(gtk::pango::EllipsizeMode::End).lines(1).width_chars(1).tooltip_text(&text).css_classes(["dim-label", "caption"]).build());
            }
        } else if self.row_highlight.get() {
            self.subtitle_box.append(&kind_subtitle_text(item, &search_subtitle(item), false));
        } else {
            self.subtitle_box.append(&kind_subtitle(item, true, false));
        }
        self.duration.set_label(&item.duration_text().unwrap_or_default());
        self.explicit.set_visible(item.explicit);
        self.dl_icon.set_visible(item.kind == ItemKind::Song && self.ctx.downloads.is_downloaded(&item.id));

        match track_number {
            Some(n) => {
                self.track_num.set_label(&n.to_string());
                self.track_num.set_visible(true);
                self.img_overlay.set_visible(false);
            }
            None => {
                self.track_num.set_visible(false);
                self.img_overlay.set_visible(true);
                match &item.thumb {
                    Some(url) => self.cover.load(url),
                    None => self.cover.set_placeholder("media-optical-symbolic"),
                }
            }
        }

        let playable = item.kind.is_playable() && !item.id.is_empty();
        self.inner.set_sensitive(playable);
        if playable {
            self.like.set_data(Some(VideoId(item.id.clone())), item.like_status);
        } else {
            self.like.set_data(None, None);
        }

        // Follow the current track. One handler per row, dropped with the row.
        if let Some(id) = self.state_handler.borrow_mut().take() {
            self.ctx.player.state().disconnect(id);
        }
        self.apply_playing(playable && self.ctx.player.state().is_playing_id(&item.id));
        if playable {
            let weak = Rc::downgrade(self);
            let video_id = item.id.clone();
            let id = self.ctx.player.state().connect_notify_local(Some("video-id"), move |state, _| {
                if let Some(row) = weak.upgrade() {
                    row.apply_playing(state.is_playing_id(&video_id));
                }
            });
            self.state_handler.replace(Some(id));
        }
    }

    fn apply_playing(self: &Rc<Self>, playing: bool) {
        let target: &gtk::Widget = if self.row_highlight.get() { self.row.upcast_ref() } else { self.inner.upcast_ref() };
        if playing {
            target.add_css_class("playing");
            target.remove_css_class("flat");
        } else {
            target.remove_css_class("playing");
            target.add_css_class("flat");
        }
        let show_indicator = playing && !self.row_highlight.get();
        self.indicator.set_visible(show_indicator);
        if show_indicator {
            self.start_animation();
        } else {
            self.stop_animation();
        }
    }

    fn start_animation(self: &Rc<Self>) {
        if self.anim.borrow().is_some() {
            return;
        }
        self.anim_state.set(false);
        let weak = Rc::downgrade(self);
        let id = glib::timeout_add_local(ANIMATION_STEP, move || {
            let Some(row) = weak.upgrade() else { return glib::ControlFlow::Break };
            let up = !row.anim_state.replace(!row.anim_state.get());
            let [a, b, c] = &row.bars;
            for (bar, raised) in [(a, up), (b, !up), (c, up)] {
                if raised {
                    bar.add_css_class("bar-up");
                } else {
                    bar.remove_css_class("bar-up");
                }
            }
            glib::ControlFlow::Continue
        });
        self.anim.replace(Some(id));
    }

    fn stop_animation(&self) {
        if let Some(id) = self.anim.borrow_mut().take() {
            id.remove();
        }
        for bar in &self.bars {
            bar.remove_css_class("bar-up");
        }
    }

    fn connect_gestures(self: &Rc<Self>) {
        let open_menu = {
            let weak = Rc::downgrade(self);
            Rc::new(move |x: f64, y: f64| {
                let Some(row) = weak.upgrade() else { return };
                let Some(track) = row.item.borrow().as_ref().and_then(MediaItem::to_track) else { return };
                let extras = row.menu_extras.borrow().clone().map(|build| build()).unwrap_or_default();
                let opts = SongMenuOptions { prefix: "row", nav: Some(row.ctx.nav.clone()), ctx: Some(row.ctx.clone()), extras, ..SongMenuOptions::default() };
                show_song_menu(&row.inner, x, y, &track, &row.ctx.player, opts);
            })
        };
        let right = gtk::GestureClick::builder().button(gdk::BUTTON_SECONDARY).build();
        let open = open_menu.clone();
        right.connect_released(move |_, _, x, y| open(x, y));
        self.inner.add_controller(right);
        let long = gtk::GestureLongPress::new();
        let open = open_menu.clone();
        long.connect_pressed(move |_, x, y| open(x, y));
        self.inner.add_controller(long);

        let left = gtk::GestureClick::builder().button(gdk::BUTTON_PRIMARY).build();
        let weak = Rc::downgrade(self);
        left.connect_pressed(move |_, _, x, y| {
            if let Some(row) = weak.upgrade() {
                row.press_at.set((x, y));
            }
        });
        let weak = Rc::downgrade(self);
        left.connect_released(move |_, _, x, y| {
            let Some(row) = weak.upgrade() else { return };
            let (sx, sy) = row.press_at.get();
            if (x - sx).abs() > CLICK_SLOP || (y - sy).abs() > CLICK_SLOP {
                return;
            }
            let handler = row.on_activate.borrow().clone();
            if let (Some(handler), Some(item)) = (handler, row.item.borrow().as_ref()) {
                handler(item);
            }
        });
        self.inner.add_controller(left);
    }
}

impl Drop for SongRow {
    fn drop(&mut self) {
        if let Some(id) = self.state_handler.borrow_mut().take() {
            self.ctx.player.state().disconnect(id);
        }
        if let Some(id) = self.anim.borrow_mut().take() {
            id.remove();
        }
    }
}
